### PR TITLE
fixes #20414 Allow different ActorMaterializer subtypes

### LINKED ISSUE #20414: Allow different ActorMaterializer subtypes
Small changes in either private[stream] class visibility or private method are required in order to allow other implementations in addition to the existing ActorMaterializerImpl.

### PR DESCRIPTION
**Pull request by kkasravi**

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-stream/src/main/scala/akka/stream/Materializer.scala [hunk 1]
```diff
@@ -73,11 +73,10 @@ private[akka] object NoMaterializer extends Materializer {
 }
 
 /**
- * INTERNAL API: this might become public later
  *
  * Context parameter to the `create` methods of sources and sinks.
  */
-private[akka] case class MaterializationContext(
+case class MaterializationContext(
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/Materializer.scala]: if opening it up, do we want to make it easier binary evolvable?

**Comment by ktoso** [comment]
Compiler not happy about missing override keyword:

```scala
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala:215: overriding method publishers in class MaterializerSession of type => java.util.Map[akka.stream.OutPort,org.reactivestreams.Publisher[Any]];
[error]  variable publishers needs `override' modifier
[error]     var publishers = Vector.empty[TestPublisher]
[error]         ^
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala:216: overriding method subscribers in class MaterializerSession of type => java.util.Map[akka.stream.InPort,AnyRef];
[error]  variable subscribers needs `override' modifier
[error]     var subscribers = Vector.empty[TestSubscriber]
[error]         ^
[error] two errors found
```

Otherwise changes look OK, let's get it though validation @kkasravi :)

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 2]
```diff
@@ -162,23 +165,16 @@ abstract class ActorMaterializer extends Materializer {
    */
   def isShutdown: Boolean
 
-  /**
-   * INTERNAL API: this might become public later
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: I am not sure about removing these markers. I don't feel that we are currently ready to make these APIs officially open. It makes sense to open up to those whom we work closely with (and hence can easily coordinate with), and also evolve it if necessary until we feel comfortable to fully commit to these APIs to the larger public.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 3]
```diff
@@ -162,23 +165,16 @@ abstract class ActorMaterializer extends Materializer {
...
-  private[akka] def system: ActorSystem
+  def actorOf(context: MaterializationContext, props: Props): ActorRef
 
-  /**
-   * INTERNAL API
-   */
-  private[akka] def logger: LoggingAdapter
+  def system: ActorSystem
+
+  def logger: LoggingAdapter
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: We have a ticket for logging in graph stage, but it will probably not be solved like this.
https://github.com/akka/akka/issues/18793

Can we keep this closed for now. You can create your own LoggingAdapter via the system, `akka.event.Logging(mat.system, classOf[Something])`

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 4]
```diff
@@ -162,23 +165,16 @@ abstract class ActorMaterializer extends Materializer {
...
 
-  /** INTERNAL API */
-  private[akka] def supervisor: ActorRef
+  def supervisor: ActorRef
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: why is this needed?

File: akka-stream/src/main/scala/akka/stream/Materializer.scala (hunk 1 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/Materializer.scala]: we can take care of that part, since it involves some boilerplate

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/Materializer.scala]: Yeah 👍

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 5]
```diff
@@ -162,23 +165,16 @@ abstract class ActorMaterializer extends Materializer {
    */
   def isShutdown: Boolean
 
-  /**
-   * INTERNAL API: this might become public later
-   */
-  private[akka] def actorOf(context: MaterializationContext, props: Props): ActorRef
+  def materialize[Mat](_runnableGraph: Graph[ClosedShape, Mat],
+                       subflowFuser: GraphInterpreterShell ⇒ ActorRef): Mat
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: GraphInterpreterShell is still in internal `impl` package. I'm not sure what it would mean to move those to public package.

An alternative would be to define a new trait `ExtendedActorMaterializer` with these methods made public. It would then live in the `impl` package (for now) and you would downcast to that instead of `ActorMaterializer`

`ActorMaterializerImpl` extends `ExtendedActorMaterializer`
`ExtendedActorMaterializer` extends `ActorMaterializer`

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: Yes, I was thinking along those lines as well. I am not sure if this is the best trait decomposition, but we can iterate on this.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala (hunk 4 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: We're not using it right now AFAIK. I'll revert.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala (hunk 5 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: Thanks @patriknw @drewhk, I'll resubmit with ExtendedActorMaterializer.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala (hunk 2 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: Sure, let me reinsert the markers so it's clear to others.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala (hunk 3 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: @patriknw ok

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 6]
```diff
@@ -163,26 +164,43 @@ abstract class ActorMaterializer extends Materializer {
   def isShutdown: Boolean
 
   /**
-   * INTERNAL API: this might become public later
+   * INTERNAL API
    */
-  private[akka] def actorOf(context: MaterializationContext, props: Props): ActorRef
+  def actorOf(context: MaterializationContext, props: Props): ActorRef
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: can this still be `private[akka]` here and public in `ExtendedActorMaterializer`?

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 7]
```diff
@@ -163,26 +164,43 @@ abstract class ActorMaterializer extends Materializer {
...
 
   /**
    * INTERNAL API
    */
-  private[akka] def system: ActorSystem
+  def system: ActorSystem
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: this should be fine, I thought that was public already

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 8]
```diff
@@ -163,26 +164,43 @@ abstract class ActorMaterializer extends Materializer {
...
+   * INTERNAL API
+   */
   private[akka] def supervisor: ActorRef
 
 }
 
 /**
+ * ExtendedActorMaterializer used by subtypes which materializer using GraphInterpreterShell
+ */
+abstract class ExtendedActorMaterializer extends ActorMaterializer {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: shall we place this in the `impl` package for now, since it also requires `GraphInterpreterShell` that is in that package?

File: akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala [hunk 9]
```diff
@@ -238,17 +234,18 @@ class StreamLayoutSpec extends AkkaSpec {
 
     materializer.subscribers.size should be(materializer.publishers.size)
 
-    val inToSubscriber: Map[InPort, TestSubscriber] = materializer.subscribers.map(s ⇒ s.port -> s).toMap
-    val outToPublisher: Map[OutPort, TestPublisher] = materializer.publishers.map(s ⇒ s.port -> s).toMap
+    import scala.collection.JavaConverters._
+    val inToSubscriber: scala.collection.mutable.Map[InPort, TestSubscriber] = materializer.subscribers.asScala.map(s ⇒ s._1 -> s._2.asInstanceOf[TestSubscriber])
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala]: why are these changes needed here?

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala (hunk 7 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: yeah, this should be fine.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 10]
```diff
@@ -745,15 +745,15 @@ private[stream] object MaterializerSession {
 /**
  * INTERNAL API
  */
-private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Module, val initialAttributes: Attributes) {
+abstract class MaterializerSession(val topLevel: StreamLayout.Module, val initialAttributes: Attributes) {
   import StreamLayout._
 
   // the contained maps store either Subscriber[Any] or VirtualPublisher, but the type system cannot express that
-  private var subscribersStack: List[ju.Map[InPort, AnyRef]] =
+  var subscribersStack: List[ju.Map[InPort, AnyRef]] =
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Hm, why make this public? The session should protect this internal state from being corrupted. I would like to hear the use-case here. Even if it turns out to be really needed by subclasses, protected might be better.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 11]
```diff
@@ -808,7 +808,7 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
     }
   }
 
-  final def materialize(): Any = {
+  def materialize(): Any = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: why remove the final?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: I think regarding the changes here, it might be better to see what is exactly the use case to see if we can factor things differently here instead of making non-final members. When dealing with inheritance I usually prefer methods/field to be either
 - private
 - public final
 - protected abstract

otherwise it becomes harder to reason about the state of the class.

**Comment by drewhk** [comment]
Getting there, but still a few things have to be thought through to avoid maintenance hell by exposing various overridable and/or public methods/state since that ties subclasses/collaborators much tighter to these classes than before in very unpredictable ways.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 10 above)
Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: @drewhk my preference was protected as well and I agree it lays bare internal state.  StreamLayoutSpec - checkMaterialized and FlatTestMaterialized drove some of the exposure after my MaterializerSession changes. Let me take another pass at this.

**Comment by patriknw** [comment]
I think this is looking good, as the first step for opening up for other materializer implementations.

**Comment by kkasravi** [comment]
@patriknw awesome - thanks.

File: akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala [hunk 12]
```diff
@@ -129,9 +130,9 @@ object ActorMaterializer {
   /**
    * INTERNAL API
    */
-  private[akka] def downcast(materializer: Materializer): ActorMaterializer =
-    materializer match { //FIXME this method is going to cause trouble for other Materializer implementations
-      case m: ActorMaterializer ⇒ m
+  def downcast(materializer: Materializer): ExtendedActorMaterializer =
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: I'm fine with opening up
`def downcast(materializer: Materializer): ActorMaterializer`
but `ExtendedActorMaterializer` should not be too visible yet. I think you can add that cast in your own code without problem.

Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: Hi @patriknw 
The problem is here 
```
private[akka] final class ActorRefSink[In](ref: ActorRef, onCompleteMessage: Any,
                                           val attributes: Attributes,
                                           shape: SinkShape[In]) extends SinkModule[In, NotUsed](shape) {

  override def create(context: MaterializationContext) = {
    val actorMaterializer = ActorMaterializer.downcast(context.materializer)
    val effectiveSettings = actorMaterializer.effectiveSettings(context.effectiveAttributes)
    val subscriberRef = actorMaterializer.actorOf(context,
      ActorRefSinkActor.props(ref, effectiveSettings.maxInputBufferSize, onCompleteMessage))
    (akka.stream.actor.ActorSubscriber[In](subscriberRef), NotUsed)
  }

  override protected def newInstance(shape: SinkShape[In]): SinkModule[In, NotUsed] =
    new ActorRefSink[In](ref, onCompleteMessage, attributes, shape)
  override def withAttributes(attr: Attributes): AtomicModule =
    new ActorRefSink[In](ref, onCompleteMessage, attr, amendShape(attr))
}
```
If we downcast to ActorMaterializer then the call to actorOf is private[akka] in the create call above.

Comment (kkasravi) [inline akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala]: Let me revert - this fails binary compatibility check.

**Comment by patriknw** [comment]
@kkasravi The downcast comment disappeared, but what I meant was that it doesn't do anything that you can't do in your source code. You can cast to `ExtendedActorMaterializer` yourself.

**Comment by kkasravi** [comment]
@patriknw yes understood, thanks. The ActorMaterializer.downcast appears in a few places that are currently internal api's. For example the ActorRefSink class has a create
```
  override def create(context: MaterializationContext) = {
    val actorMaterializer = ActorMaterializer.downcast(context.materializer)
    val effectiveSettings = actorMaterializer.effectiveSettings(context.effectiveAttributes)
    val subscriberRef = actorMaterializer.actorOf(context,
      ActorRefSinkActor.props(ref, effectiveSettings.maxInputBufferSize, onCompleteMessage))
    (akka.stream.actor.ActorSubscriber[In](subscriberRef), NotUsed)
  }
```

**Comment by drewhk** [comment]
LGTM from my side

**Comment by patriknw** [comment]
@kkasravi looks like this requires a rebase, then we can merge (sorry for the inconvenience)

**Comment by drewhk** [comment]
The downcast method was just a convenience internal method for us. Maybe best is just to remove it and add the explicit casts where the method was used (i.e. inline it).

**Comment by kkasravi** [comment]
@patriknw Sure, will rebase. 
@drewhk Yes, I'll do the downcasts explicitly. 

I'll push the PR for 'GraphStage subtypes should not be private to akka #20543' once this is merged since it includes these changes.

**Comment by kkasravi** [comment]
@drewhk 
Choice of throwing ClassCastException or IllegalArgumentException
```
materializer.asInstanceOf[ActorMaterializer]
```
vs
```
materializer match {
    case m: ActorMaterializer ⇒ m
    case _ ⇒ throw new IllegalArgumentException(s"required [${classOf[ActorMaterializer].getName}] " +
        s"but got [${materializer.getClass.getName}]")
}
```

I went with the hard cast, I didn't think the information provided in IllegalArgumentException provided much more information, but i will go with your recommendation.

**Comment by drewhk** [comment]
I like the explicit exception. Maybe the best option is to extract this method into a separate object that is not public API.

**Comment by kkasravi** [comment]
@drewhk ah ok - maybe something like:
```
object ActorMaterializerHelper {
  /**
   * INTERNAL API
   */
  def downcast(materializer: Materializer): ActorMaterializer =
    materializer match { //FIXME this method is going to cause trouble for other Materializer implementations
      case m: ActorMaterializer ⇒ m
      case _ ⇒ throw new IllegalArgumentException(s"required [${classOf[ActorMaterializer].getName}] " +
        s"but got [${materializer.getClass.getName}]")
    }
}
```

**Comment by kkasravi** [comment]
@drewhk or
```
/**
 * INTERNAL API
 */
object ActorMaterializerHelper {
  def downcast(materializer: Materializer): ActorMaterializer =
    materializer match { //FIXME this method is going to cause trouble for other Materializer implementations
      case m: ActorMaterializer ⇒ m
      case _ ⇒ throw new IllegalArgumentException(s"required [${classOf[ActorMaterializer].getName}] " +
        s"but got [${materializer.getClass.getName}]")
    }
}
```

**Comment by kkasravi** [comment]
@drewhk @patriknw 
binary compatibility failure
```
[info] akka-stream: found 1 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.2  (filtered 193)
[error]  * method downcast(akka.stream.Materializer)akka.stream.ActorMaterializer in object akka.stream.ActorMaterializer does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.ActorMaterializer.downcast")
```
I thought this would be ignored if marked Internal. I guess not. @drewhk did you want the separate object with the downcast method as well as the downcast kept in ActorMaterializer for backward compatibility reasons?

**Comment by patriknw** [comment]
@kkasravi sorry that I missed that question. You need to add that filter to [Mima.scala](https://github.com/akka/akka/blob/master/project/MiMa.scala#L878)

```
ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.ActorMaterializer.downcast")
```

**Comment by kkasravi** [comment]
@patriknw awesome, thanks! I'll resubmit this PR then.

**Comment by kkasravi** [comment]
@patriknw @drewhk PR is ready to go. I have a followup PR for #20543 once this is merged.
thanks

**Comment by patriknw** [comment]
@drewhk ok that I merge this now?

**Comment by drewhk** [comment]
Yeah merge it!
